Add additional bridge port settings

chdxD1 opened the pull request:
## Description
This MR adds additional bridge settings `hairpin` and `port-mac-learning` on bridge ports. They are useful when using VXLAN tunnels with e.g. FRR.

This is my first MR to netplan, I tried my best to run the tests locally first.

## Checklist

- [x] Runs `make check` successfully. (except some nmcli local tests)
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented. (but with version ???)
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

chdxD1 commented:
Also I am not sure regarding the ABI re-generate, it seems there are quite a lot of changes if done on this branch which are not related?

daniloegea reviewed (changes requested):
Hi, thank you for your PR it's much appreciated.

I started reviewing it from the first commit but then realized you addresses most of the problems I pointed out and removed the `mcast-snooping` part in the last commit.

May I ask you to squash your commits and update the PR? It'll be easier to review.

Thanks again!

chdxD1 commented:
Yeah, mcast-snooping was a bit of a different topic (especially it needs to be default on, like STP), we will be using a systemd override for that internally for the time being.

Same goes for vlan_filtering on the bridge, I've seen the VLAN MR (#127) and it probably doesn't make sense to implement an option to *just* configure vlan_filtering without the backend actually configuring some. (We have this very specific case, backend just needs to set this flag, VLANs are then set by some K8s operator)

Anyway, I've squashed the trial/error commits (sorry for the additional lost review work there) and adjusted it according to the review comments :)

daniloegea commented:
Hi Christopher, sorry for the delay.

It's looking good. I created a couple of integration tests (and executed them with autopkgtest) using these two new options. Feel free to integrate them into your PR: https://github.com/daniloegea/netplan/commit/e543ddcecfef6e4167e2ca59acf12b1e71f17f25

As it's adding two new keys to Netplan, I'll ask @slyon what he thinks.

chdxD1 commented:
@daniloegea I finally found the time to add your commit, thank you for the test cases!

slyon commented on doc/netplan-yaml.md:
```diff
@@ -1327,6 +1339,12 @@ The specific settings for bridges are defined below.
 
+  - **mcast-snooping** (bool)
+
+    > Define whether the bridge should use multicast sooping. The
```
The `mcast-snooping` logic was dropped from this PR.

slyon commented on doc/netplan-yaml.md:
```diff
@@ -1327,6 +1339,12 @@ The specific settings for bridges are defined below.
+  - **mcast-snooping** (bool)
+
+    > Define whether the bridge should use multicast sooping. The
+    > default value is "true", which means that multicast snooping
```
The `mcast-snooping` logic was dropped from this PR.

slyon commented on doc/netplan-yaml.md:
```diff
@@ -591,7 +591,20 @@ Match devices by MAC when setting options like: `wakeonlan` or `*-offload`.
+  > Takes a boolean. Configures whether traffic may be sent back out of the
+  > bridge port on which it was received. When this flag is false, then the
+  > bridge will not forward traffic back out of the receiving port. When
+  > unset, the kernel's default will be used.
```
todo: "When unset, the **backend's** default will be used."

(This is FALSE for NetworkManager, and the kernel's value for systemd-netword)

slyon commented on src/parse.c:
```diff
@@ -2807,6 +2807,8 @@ static const mapping_entry_handler dhcp6_overrides_handlers[] = {
     {"routes", YAML_SEQUENCE_NODE, {.generic=handle_routes}, NULL}, \
     {"routing-policy", YAML_SEQUENCE_NODE, {.generic=handle_ip_rules}, NULL}, \
+    {"hairpin", YAML_SCALAR_NODE, {.generic=handle_netdef_tristate}, netdef_offset(bridge_hairpin)}, \
+    {"learning", YAML_SCALAR_NODE, {.generic=handle_netdef_tristate}, netdef_offset(bridge_learning)}, \
```
```suggestion
    {"mac-learning", YAML_SCALAR_NODE, {.generic=handle_netdef_tristate}, netdef_offset(bridge_learning)}, \
```

todo: I feel like we should call this mac-learning, to stay in line with the corresponding VXLAN setting of similar functionality.

slyon commented on src/networkd.c:
```diff
@@ -124,6 +124,7 @@ write_bridge_params_networkd(GString* s, const NetplanNetDefinition* def)
         if (def->bridge_params.max_age)
             g_string_append_printf(params, "MaxAgeSec=%s\n", def->bridge_params.max_age);
         g_string_append_printf(params, "STP=%s\n", def->bridge_params.stp ? "true" : "false");
+        g_string_append_printf(params, "MulticastSnooping=%s\n", def->bridge_params.stp ? "true" : "false");
```
note: mcast_snooping has been dropped from this PR.

slyon reviewed (commented):
Thank you @chdxD1 for your contribution to Netplan, and sorry for taking so long to comment.

I like the overall shape of this PR, thanks @daniloegea for your initial review! Those comments have been addressed already. Tests and CI are looking good, too! Thanks for adding/integrating unit- and integration tests.

I left some small inline-comments, mostly about documentation and naming of the new YAML schema.

We're introducing two new YAML settings here (let me try to confirm the schema with our architect):
* **hairpin** => We should **keep this as suggested**, IMO. "hairpin" is descriptive and common wording.
  * called "HairPin" in networkd
  * called "hairpin-mode" in NetworkManager
  * called "hairpin_mode" in iproute2
  * called "reflective relay" in Cisco/Juniper
* **learning** => We should **change this to "mac-learning"**, IMO. Just "learning" might be a bit too generic. Also, we should stay in-line with the "mac-learning" setting for VXLAN devices.
  * called "Learning" in systemd-networkd
  * Not generically implemented in NetworkManager (only for VXLAN)
  * called "learning" in iproute2
  * called "MAC Address Source-based Learning" in Cisco

The "learning" => "mac-learning" suggestion brings up one issue: What happens if a VXLAN device is used as the bridge-port? That way the new "mac-learning" setting would probably trump the VXLAN "mac-learning" setting and lead to unexpected behavior. You mention VXLAN tunnels in combination with FRR in your PR description. Could you please provide some more context on that? Would it be useful to actually re-use the VXLAN "mac-learning" setting and refactor it to become a generic setting. We could then write the "MacLearning=" setting in .netdev file and "Learning=" in .network file (for systemd-networkd), skipping the .netdev part on non-VXLAN ports. NetworkManager would not be affected, as it only supports the vxlan.learning setting (not on bridge-ports).

PS: rebasing your branch onto `origin/main` should fix the "RPM build (fedora:latest)" CI. The other failures are known false-positives.

vorlonofportland reviewed (approved):
These key names are straightforward and consistent with naming in upstream tooling, +1 for the schema changes.

vorlonofportland commented:
To be clear I'm ok with either 'learning' or 'port-mac-learning' as a name.

slyon reviewed (approved):
Thanks for the schema approval. Let's go with port-mac-learning then for my reasons given above.
I've rebased and pushed the corresponding renaming changes.

I think this should be ready for merging, just in time for the next Netplan release :)
This should cover some more cases for https://github.com/telekom/netplanner
